docs/react-types/ComponentProps.md

```diff
@@ -0,0 +1,76 @@
+---
+title: ComponentProps<ElementType>
+---
+
+`ComponentProps<ElementType>` constructs a type with all valid props of an element or inferred props of a component.
+
+:::note
+
+Prefer `ComponentPropsWithRef<ElementType>` if ref is forwarded and `ComponentPropsWithoutRef<ElementType>` when ref is not forwarded.
```

Comment:
I personally always found that confusing. The whole point of `ComponentProps` is to get the props but at the same time I should know that it has a `ref`? Or does that mean `ComponentPropsWithoutRef ` is just a shorthand for `Omit<ComponentPropsWithRef<T>, 'ref'>`?

Either way, this confusion already exist in the types so I'd hope we can do a follow-up with either clearer instructions or a plan to only `ComponentProps` and deprecate/remove the ones related to `ref`.

Reply:
Agreed, I tend to use `ComponentProps` despite the recommendation to prefer "being explicit about the ref", and I have never had problems with it.

I made a small experiment and found that `ComponentProps` probably is a better default when passing a component type, because it correctly infers whether or not ref is forwarded. 

[TypeScript playground](https://www.typescriptlang.org/play?jsx=4#code/JYWwDg9gTgLgBAbzgYQuCA7AphmAFKCMAZwBoU1JtcCjiB1YGACwCUsAzc1da-QkoxYQArjHZc4HaAHcAhlAAmEuAF8phEHABEULHIDGMbQCgTwXFigdDWOLRKITcODCwAPGAC44xGFAsAcxNVMwNMPwpeHBghNk44AF4pWQVlTgAeAAkAFQBZABkAEWAANwBRABssEBjyB2IAPgAKDhEMI2BMKKoYuIlm50RXD3hVUiG9LhMASicXPRgRKAw4DMUyuCnEhCnVRoQ3T1UMgHoN0saQmbM2jpgu1Z5e3DjRcU5mpCOxnwa5hCTLBLFZrC4HH4nc5lK6hEwwACeYDsAEEkj1MDEGhlEciIBwMXx+pwrqdTi44AA9AD8ZlxdgAQujnpiaAJiDikVh8YS+kxmO8JKTyS4aXSuShmZRWfw6MSOJy8QSWUT+UKTGSKWL4RKilLomy5WrMvSeSq+cIxOrNaLaTrkXByvqXrLBPzBSauWbpaqWNaRVS7fS4AAxZ0yhpvK2epW817u6McYVau2nABUzjTcAABubDSRFdzlT6LfEk9m4MBiHAsABHERlOTVXCuCA5vOuhjGhWm4sGzvyxrZzNwEe5kv5jm9uOxBMfcuV6t1hulJsxVvticDucSQve-tR+dDkcjgCSqwARhAWHADHJiFgyJv+w0K1W4HI4BfgW4oHBFJwcgiJU8DEBYBh2EwcAyMAlSVLe0B6EYlQIpWGAcFY0HMMC2F-tAcAYNeWwJO+0hQPIShYIoAB0manEAA)

Is there a case for being explicit when passing an element string literal? Is `ComponentPropsWithRef<"div">` even something you'd want to do?

Reply:
I prefer having a single `ComponentProps` if the only justification for `ComponentPropsWithoutRef` is a shorthand for `ComponentProps` + `Omit`.

Long-term we won't have `forwardRef` anyway and then `ref` will be in the props interface for function components.

Reply:
Do you have any insights into how you'll define `ref`s in the future? Just a regular prop like this?

```tsx
interface Props {
  ref: Ref
}

function Component({ ref }: Props) {
  return <div ref={ref} />
}
```

Reply:
I'd like to investigate if we can deprecate `ComponentPropsWithRef` and `ComponentPropsWithoutRef` in @types/react. Any tips on how to go about it? Make a PR with these thoughts? Someone in particular to get in touch with before taking that step?

Reply:
I'd try to find out when they got introduced and why. Then deprecate and provide a counter argument to the initial rationale. By default we don't want to add stuff unless we have to. And if it is convenience there needs to be frequent usage.